core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -199,17 +204,27 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
             IOException lastException = null;
             int maxLocalStorageNodes = MAX_LOCAL_STORAGE_NODES_SETTING.get(settings);
             for (int possibleLockId = 0; possibleLockId < maxLocalStorageNodes; possibleLockId++) {
-                for (int dirIndex = 0; dirIndex < environment.dataFiles().length; dirIndex++) {
-                    Path dataDirWithClusterName = environment.dataWithClusterFiles()[dirIndex];
-                    Path dataDir = environment.dataFiles()[dirIndex];
+                for (int dirIndex = 0; dirIndex < environment.dataFiles().length + extra; dirIndex++) {
+                    final Path dataDir;
+                    if (dirIndex < environment.dataFiles().length) {
```

Comment:
I completely agree with deprecating multiple nodes on the same data directory. I think that we can do that as a follow-up?